bpf/bpf.go

```diff
@@ -1024,6 +1025,59 @@ func (b *BPFLib) getMapArgs(ifName string) ([]string, error) {
 	return mapArgs, nil
 }
 
+func (b *BPFLib) getCalicoBPFProgIds() ([]int, error) {
+	// Get a list of ids of BPF programs that loaded by Calico.
+	// Programs are identified by 'name', inherited from func name.
+	// for example, in bpf/bpf-apache/filter.c, a function name is 'prefilter'.
+	progName := "prefilter"
+
+	prog := "bpftool"
+	args := []string {
+		"--json",
+		"--pretty",
+		"prog",
+		"show",
+	}
+	printCommand(prog, args...)
+	output, err := exec.Command(prog, args...).CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of BPF programs: %s\n%s", err, output)
+	}
+	var p []progInfo
+	err = json.Unmarshal(output, &p)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse json output: %v\n%s", err, output)
+	}
+
+	var id = []int{}
+	for _, v := range p {
+		if v.Name == progName {
+			id = append(id, v.Id)
+		}
+	}
+
+	return id, nil
+}
+
+func (b *BPFLib) canRemoveXDP(ifName string, id []int) (bool, error) {
+	i, err := b.GetXDPID(ifName)
+	if err != nil {
+		// iproute2 error or Empty
+		if i == -1 {
```

Comment:
Please comment here to say what this case means.

Reply:
Originally, GetXDPID() returns -1 regardless of the type of error. But in my implementation I need to know why iproute2 returned an error.
This is because, even if GetXDPID() does not find the id (=even if the XDP program is not loaded), the RemoveXDP() function needs to be executed in the resync period.

I also think this method is not neat. Maybe it would be better to separate a function that checks for the existence of an XDP program.
I want to hear from you.

Reply:
Sorry, still understanding some detail here.  Why is `RemoveXDP()` needed if there isn't any XDP program attached to the interface?

Reply:
Ah, now I understand what you said.
If the interface does not have an XDP program, fixupXDPProgramAndMapConsistency() does not add UninstallXDP, and RemoveXDP() is not executed in XDPAction.apply().

Therefore, it is unnecessary to check whether the XDP program exists in GetXDPID().
Am I right?